relay-replays/src/serialization.rs

```diff
@@ -0,0 +1,411 @@
+use serde::de::Visitor;
+use serde::{Deserialize, Deserializer, Serialize};
+use serde_json::Value;
+
+use crate::recording::*;
+
+/// Implementation tweaked from serde's `derive(Deserialize)` for internally tagged enums,
+/// in order to work with integer tags.
+impl<'de> Deserialize<'de> for Event {
+    fn deserialize<D>(d: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        enum EventType {
+            T0,
+            T1,
+            T2,
+            T3,
+            T4,
+            T5,
+            T6,
+        }
+        struct EventTypeVisitor;
+        impl<'de> Visitor<'de> for EventTypeVisitor {
+            type Value = EventType;
+            fn expecting(&self, formatter: &mut ::std::fmt::Formatter) -> ::std::fmt::Result {
+                ::std::fmt::Formatter::write_str(formatter, "type id")
+            }
+            fn visit_u64<E>(self, value: u64) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                use EventType::*;
+                match value {
+                    0 => Ok(T0),
+                    1 => Ok(T1),
+                    2 => Ok(T2),
+                    3 => Ok(T3),
+                    4 => Ok(T4),
+                    5 => Ok(T5),
+                    6 => Ok(T6),
+                    _ => Err(serde::de::Error::invalid_value(
+                        serde::de::Unexpected::Unsigned(value),
+                        &"type id 0 <= i < 7",
+                    )),
+                }
+            }
+        }
+
+        impl<'de> serde::Deserialize<'de> for EventType {
+            #[inline]
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: Deserializer<'de>,
+            {
+                Deserializer::deserialize_any(deserializer, EventTypeVisitor)
+            }
+        }
+        let tagged = match Deserializer::deserialize_any(
+            d,
+            serde::__private::de::TaggedContentVisitor::<EventType>::new(
+                "type",
+                "internally tagged enum Event",
+            ),
+        ) {
+            Ok(val) => val,
+            Err(err) => {
+                return Err(dbg!(err));
+            }
+        };
+        let content_deserializer =
+            serde::__private::de::ContentDeserializer::<D::Error>::new(tagged.content);
+        match tagged.tag {
+            EventType::T0 => Result::map(
+                <Value as serde::Deserialize>::deserialize(content_deserializer),
+                Event::T0,
+            ),
```

Comment:
Did you have that code expand/generate somewhere? It reads like machine-written or macro code.

```suggestion
            EventType::T0 => Value::deserialize(content_deserializer).map(Event::T0),
```

Reply:
Yes, I basically copied the `cargo expand` of an internally tagged enum and tweaked it to work with integer tags. That is also where the `EventType` indirection comes from. Will simplify.